core/src/main/scala/spinal/core/internals/Phase.scala

```diff
@@ -898,7 +898,26 @@ class PhaseMemBlackBoxingDefault(policy: MemBlackboxingPolicy) extends PhaseMemB
     }
 
     if (mem.initialContent != null) {
-      return "Can't blackbox ROM"  //TODO
+      if (topo.portCount == 1 && topo.readsSync.size == 1 && topo.readWriteSync.isEmpty) {
+        mem.component.rework {
+          val port = topo.readsSync.head
+
+          val rom = new Rom_1rs(
+            wordWidth = port.width,
+            wordCount = mem.wordCount*mem.width/port.width,
+            technology = mem.technology
+          )
+
+          rom.io.addr.assignFrom(port.address)
+          rom.io.en.assignFrom(port.clockDomain.isClockEnableActive)
+
+          wrapConsumers(port, rom.io.data)
+
+          rom.setName(mem.getName())
```

Comment:
I've got a question about how to approach a problem I'm having refactoring this.

I understand that there should be a `removeMem()` call right around here.

However, when I do that, `emitMem()` is no longer called. 

In order to generate `emitMem()`, I need the phase context and some other state in order to figure out the name of the file and directory to put the black box in. 

I think the conflict lies in this difference:
- black box RAM models are meant to be static files provided by the user, and thus no generation is required
- for the black box ROM model, I'd like it to be generated by the system, because the ROM contents are actually inside SpinalHDL.
- Furthermore, I'd like the black box itself to contain a `case` statement, so that if there is a failure to replace the blackbox model with an actual hardware macro, in the worst case the ROM is tossed directly into the synthesizer and it's generated "as gates" - in a modern process this might be preferable for ROMs that are less than a few hundred bytes in size. Either way, you at least get a "correct" circuit even if the back-end team makes a mistake and fails to replace the black box with a hard macro file.

Can you suggest how I could generate the ROM file at this point in the output compilation phase?

Reply:
Hi,

> However, when I do that, emitMem() is no longer called.

Yes that is on purpose.
The idea of the SpinalHDL blackboxing  phase is to only emit it as a blackbox durring the generation phase, so SpinalHDL will only see a blackbox in the future, no more Mem => no emitMem

> black box RAM models are meant to be static files provided by the user, and thus no generation is required

Right. What i did a few time, is to add another SpinalHDL transformation phase, which will look at all the emited memory blackbox and transform them into something more specific (for ASIC), for instance : 
- automaticaly adding some test IO to the blackbox
- changing their name to depend on their parameter, for instance transforming the name Ram_1w_1rs into Ram_1w_1rs_1024x32, aswell as generating scripts for openram to generate them.

> for the black box ROM model, I'd like it to be generated by the system, because the ROM contents are actually inside SpinalHDL. 

I think here a custom blackboxing phase could be added to either generate a verilog of that blackboxed rom, or either export the rom content into some binary file for other usages.

Here is an example : 
https://github.com/SpinalHDL/SpinalHDL/commit/1d86c31bfa90d6c328e6498d264f8c97221ed8ee
It is based on your branch.

That will generate for each rom a verilog file with things as : 

```verilog
module xy_rom (
  input  wire clk,
  input  wire en,
  output wire [7:0] addr,
  input  wire [31:0] data
);
  always @(posedge clk) begin
    if(en) begin
      case (address)
        8'h0: data <= 32'hff
        8'h1: data <= 32'hfe
        8'h2: data <= 32'hfd
...
        8'hfc: data <= 32'h3
        8'hfd: data <= 32'h2
        8'hfe: data <= 32'h1
        8'hff: data <= 32'h0
      endcase
    end
  end
endmodule
```

Also, carefull your branch miss this : 
https://github.com/SpinalHDL/SpinalHDL/commit/e57f89dfb90b40cc13306418197a3000d88eaef7

Next week i will be in travel / holiday, so away from the keyboard :)

Reply:
I *think* I vaguely follow the approach you're suggesting. I will need to poke at it a bit more.

> Next week i will be in travel / holiday, so away from the keyboard :)

No worries, same here as well. I don't think I'll be able to make the changes so soon, but also, this isn't urgent (I've got a tape-out candidate generated from my own branch that passes the tests I need).

I'm mostly just trying to merge the changes as an exercise to get me more comfortable with SpinalHDL, while hopefully adding a minor feature to the code base. Hopefully in the future I can do more meaningful contributions.

I appreciate the review and guidance, we can pick this up again after you're back from holiday!

Reply:
> I'm mostly just trying to merge the changes as an exercise to get me more comfortable with SpinalHDL, while hopefully adding a minor feature to the code base. Hopefully in the future I can do more meaningful contributions. 

No worries ^^

> I appreciate the review and guidance, we can pick this up again after you're back from holiday!

Yes :)